src/main/haskell/kore/src/Kore/Step/BaseStep.hs

```diff
@@ -170,6 +179,18 @@ instance
     freshVariableWith (ConfigurationVariable a) n =
         ConfigurationVariable $ freshVariableWith a n
 
+{-! The result of applying an axiom to a pattern. Contains the rewritten
+pattern (if any) and the unrewritten part of the original pattern.
+-}
+data StepResult level variable =
```

Comment:
Is this not `OrOfExpandedPattern` by another name, and with only two disjuncts?

Reply:
Well, maybe, but it's tricky. First, @traiansf uses this formula in https://github.com/kframework/kore/pull/268 :
```
(∀ X . Φ'(X) → ◆ ∃ Y . ψ(X, Y))
∧
(∀ X . Φ(X) ∧ (¬ ∃ Z. α(Z)) → ⚪ ◆ ∃ Y . ψ(X, Y))
```
I believe that "next" can be removed from the last term, i.e. we have something like
```
(∀ X . Φ'(X) → ◆ ∃ Y . ψ(X, Y))
∧
(∀ X . Φ(X) ∧ (¬ ∃ Z. α(Z)) →  ◆ ∃ Y . ψ(X, Y))
```
in which case this becomes
```
(
    (∀ X . Φ'(X))
    ∨ 
    (∀ X . Φ(X) ∧ (¬ ∃ Z. α(Z)))
) → ◆ ∃ Y . ψ(X, Y))
```
in which case, indeed, it becomes an `or`. I'll wait to see what @traiansf does with his PR before agreeing that this is, indeed, the case. We may have a small issue because the `Φ` formula above is not what we are actually rewriting, but whatever.

Anyway, we have a second issue: the remainder might be part of an `or`, but we might want to treat it as an `or` only when we do proofs, not when we execute. To be more precise, I guess that when we execute a step we always want to rewrite the initial pattern, and we may want to ignore the reminders (or, at most, we may to rewrite the initial pattern in all possible ways, then we may want to generate a reminder for all applied rules at once). Again, @traiansf to confirm this.

Reply:
Things have become a bit more clear about the first issue mentioned in the above message. Here's how I understand this problem:

This rewriting stuff is useful when we want to do single-path execution and we want to make sure we reach a certain target. Roughly, we want to first remove the target from our current pattern, then we want to rewrite the result entirely. 

There are two ways of doing this. The first one is described in Traian's PR mentioned above, and there we want to treat the remainder separately, rewriting it, and rewriting the new remainder, and so on, until we get a bottom remainder. Failing to do so means failing to reach the target.

In the second one, we would represent the pattern as an "or" and we would take random components (`rc`) of that or, removing the target (`t`) from them (`rt := rc and not (exists x . t)`). If `rt` is null, then we just remove `rc` from the or and we restart the algorithm. If `rt` is not null we would try to an axiom `lhs->o rhs` axiom with which to rewrite `rt`. If we can't find one, we fail the algorithm. If we find one, then we remove `rc` from the or and we add the rewrite result together with the `rc` from which we removed the axiom `lhs` (`rc and not (exists x . lhs)`) back to the or. We repeat everything until the or becomes empty.

My guess is that @traiansf prefers the former, but not the latter.

The second issue from above still remains unsolved.